### PR TITLE
snek-restart console script

### LINKED ISSUE #173: Command line snek-restart
Similar to fluidsim-restart but adapted for snek5000 (see https://github.com/snek5000/snek5000-cbox/blob/main/doc/examples/nonlinear_restart/run_restart.py).

```
$ fluidsim-restart -h
usage: fluidsim-restart [-h] [--t_approx T_APPROX] [--merge-missing-params] [-oc] [-oi] [--new-dir-results] [--add-to-t_end ADD_TO_T_END]
                        [--add-to-it_end ADD_TO_IT_END] [--t_end T_END] [--it_end IT_END] [--modify-params MODIFY_PARAMS] [--max-elapsed MAX_ELAPSED]
                        path

Restart a fluidsim simulation

positional arguments:
  path                  Path of the directory or file from which to restart

options:
  -h, --help            show this help message and exit
  --t_approx T_APPROX   Approximate time to choose the file from which to restart
  --merge-missing-params
                        Can be used to load old simulations carried out with an old fluidsim version.
  -oc, --only-check     Only check what should be done
  -oi, --only-init      Only run initialization phase
  --new-dir-results     Create a new directory for the new simulation
  --add-to-t_end ADD_TO_T_END
                        Time added to params.time_stepping.t_end
  --add-to-it_end ADD_TO_IT_END
                        Number of steps added to params.time_stepping.it_end
  --t_end T_END         params.time_stepping.t_end
  --it_end IT_END       params.time_stepping.it_end
  --modify-params MODIFY_PARAMS
                        Code modifying the `params` object.
  --max-elapsed MAX_ELAPSED
                        Maximum elapsed time.
```

See the fluidsim implementation:
- https://foss.heptapod.net/fluiddyn/fluidsim/-/blob/branch/default/setup.cfg#L61
- https://foss.heptapod.net/fluiddyn/fluidsim/-/blob/branch/default/fluidsim/util/scripts/restart.py

### PR DESCRIPTION
**Pull request by paugier**
Fixes #173

Needs https://foss.heptapod.net/fluiddyn/fluidsim/-/merge_requests/320

@ashwinvis I think I'm going to need you help on this one :-)

### CODE REVIEW

**Comment by paugier** [comment]
For now, the help is:

```
$ snek-restart -h
usage: snek-restart [-h] [--t_approx T_APPROX] [-oc] [-oi] [--new-dir-results] [--add-to-t_end ADD_TO_T_END] [--add-to-it_end ADD_TO_IT_END] [--t_end T_END] [--it_end IT_END]
                    [--modify-params MODIFY_PARAMS] [-np NB_MPI_PROCS] [--use-start-from USE_START_FROM] [--use-checkpoint USE_CHECKPOINT] [--session-id SESSION_ID]
                    [--skip-verify-contents]
                    path

Restart a fluidsim simulation

positional arguments:
  path                  Path of the directory or file from which to restart

options:
  -h, --help            show this help message and exit
  --t_approx T_APPROX   Approximate time to choose the file from which to restart
  -oc, --only-check     Only check what should be done
  -oi, --only-init      Only run initialization phase
  --new-dir-results     Create a new directory for the new simulation
  --add-to-t_end ADD_TO_T_END
                        Time added to params.time_stepping.t_end
  --add-to-it_end ADD_TO_IT_END
                        Number of steps added to params.time_stepping.it_end
  --t_end T_END         params.time_stepping.t_end
  --it_end IT_END       params.time_stepping.it_end
  --modify-params MODIFY_PARAMS
                        Code modifying the `params` object.
  -np NB_MPI_PROCS, --nb-mpi-procs NB_MPI_PROCS
                        Number of MPI processes
  --use-start-from USE_START_FROM
                        Name of the field file to restart from. Mutually exclusive option with `use_checkpoint`.
  --use-checkpoint USE_CHECKPOINT
                        Number of the multi-file checkpoint file set to restart from. Mutually exclusive parameter with `use_start_from`.
  --session-id SESSION_ID
                        Indicate which session directory should be used to look for restart files. If not specified it would default to the `path_session` value last recorded in
                        the `params_simul.xml` file.
  --skip-verify-contents
                        Do not verify directory contents to avoid runtime errors.
```

`--t_approx` is not taken into account. I'm not sure how it should be used.

I don't know if  `--new-dir-results` could be obtained with snek5000?

**Comment by paugier** [comment]
I'm going to do again a small change in fluidsim_core and here to have something more coherent with Nek/Snek for the time parameters.

**Comment by paugier** [comment]
The help message is now:

```
$ snek-restart -h
usage: snek-restart [-h] [-oc] [-oi] [--modify-params MODIFY_PARAMS] [-np NB_MPI_PROCS] [--use-start-from USE_START_FROM]
                    [--use-checkpoint USE_CHECKPOINT] [--session-id SESSION_ID] [--skip-verify-contents]
                    [--add-to-end-time ADD_TO_END_TIME] [--end-time END_TIME] [--num-steps NUM_STEPS]
                    path

Restart a fluidsim simulation

positional arguments:
  path                  Path of the directory or file from which to restart

options:
  -h, --help            show this help message and exit
  -oc, --only-check     Only check what should be done
  -oi, --only-init      Only run initialization phase
  --modify-params MODIFY_PARAMS
                        Code modifying the `params` object.
  -np NB_MPI_PROCS, --nb-mpi-procs NB_MPI_PROCS
                        Number of MPI processes
  --use-start-from USE_START_FROM
                        Name of the field file to restart from. Mutually exclusive option with `use_checkpoint`.
  --use-checkpoint USE_CHECKPOINT
                        Number of the multi-file checkpoint file set to restart from. Mutually exclusive parameter with
                        `use_start_from`.
  --session-id SESSION_ID
                        Indicate which session directory should be used to look for restart files. If not specified it would default to
                        the `path_session` value last recorded in the `params_simul.xml` file.
  --skip-verify-contents
                        Do not verify directory contents to avoid runtime errors.
  --add-to-end-time ADD_TO_END_TIME
                        Time added to params.nek.general.end_time
  --end-time END_TIME   params.nek.general.end_time
  --num-steps NUM_STEPS
                        params.nek.general.num_steps
```

@ashwinvis I think you can now have a look at this restart command and fix what needs to be fixed.

**Comment by paugier** [comment]
I need to release fluidsim 0.7.0 for something else. So it's going to be too late to modify fluidsim-core for this PR. However, we can have a new fluidsim-core release before Snek5000 0.9.0.

File: src/snek5000/util/restart.py [hunk 1]
```diff
@@ -262,21 +269,182 @@ def load_for_restart(
...
+
+def main():
+    _restarter.restart()
+
+
+if "sphinx" in sys.modules:
+    from textwrap import indent
+    from unittest.mock import patch
+
+    with patch.object(sys, "argv", ["fluidsim-restart"]):
```
Comment (ashwinvis) [inline src/snek5000/util/restart.py:440]:
```suggestion
    with patch.object(sys, "argv", ["snek-restart"]):
```

File: src/snek5000/util/restart.py [hunk 2]
```diff
@@ -262,21 +269,182 @@ def load_for_restart(
...
+        if sum(arg is not None for arg in args_times) > 1:
+            raise ValueError(
+                "--add-to-end-time, --end-time and --num-steps are exclusive options."
+            )
+
+        if params.nek.general.stop_at == "numSteps" and args.num_steps is None:
+            raise ValueError(
+                'When params.nek.general.stop_at == "numSteps", '
+                "--num-steps should be given."
+            )
```
Comment (ashwinvis) [inline src/snek5000/util/restart.py:416]: Is this check necessary? We can override how the simulation stops in the restart run, can't we?

Anyways we are ensuring above:

```py
        if args.num_steps is not None:
            params.nek.general.stop_at = "numSteps"
```

File: tests/test_restart_command.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        "snek-restart",
+        sim_cbox_executed.output.path_run,
+        "--use-checkpoint",
+        "1",
+        "--new-dir-results",
+        "--add-to-end-time",
+        "0.002",
+    ]
+    with patch.object(sys, "argv", command):
+        main()
```
Comment (ashwinvis) [inline tests/test_restart_command.py:67]: Can we load the restart run and see if it actually worked?

**Review by ashwinvis — CHANGES_REQUESTED** [review]

File: src/snek5000/util/restart.py (hunk 2 above)
Comment (paugier) [reply src/snek5000/util/restart.py]: i removed the check which was not very useful and potentially problematic.

File: tests/test_restart_command.py (hunk 3 above)
Comment (paugier) [reply tests/test_restart_command.py:95]: We now have much better testing. And of course, there was some (evident) bugs!

**Pull request merged**